File: step_counter.go
// step_counter.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"sort"
	"strconv"
	"time"
)

const dataFile = "steps.json"
const defaultGoal = 10000

type HistoryEntry struct {
	Date  string `json:"date"`
	Steps int    `json:"steps"`
}

type Data struct {
	Goal    int            `json:"goal"`
	History []HistoryEntry `json:"history"`
}

type Counter struct {
	data Data
}

func (c *Counter) load() {
	file, err := os.ReadFile(dataFile)
	if err != nil {
		c.data = Data{Goal: defaultGoal, History: []HistoryEntry{}}
		return
	}
	if err := json.Unmarshal(file, &c.data); err != nil {
		c.data = Data{Goal: defaultGoal, History: []HistoryEntry{}}
	}
}

func (c *Counter) save() {
	data, _ := json.MarshalIndent(c.data, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (c *Counter) getToday() *HistoryEntry {
	today := time.Now().Format("2006-01-02")
	for i := range c.data.History {
		if c.data.History[i].Date == today {
			return &c.data.History[i]
		}
	}
	return nil
}

func (c *Counter) addSteps(steps int) {
	today := time.Now().Format("2006-01-02")
	entry := c.getToday()
	if entry != nil {
		entry.Steps += steps
	} else {
		c.data.History = append(c.data.History, HistoryEntry{Date: today, Steps: steps})
	}
	c.save()
	total := 0
	if entry != nil {
		total = entry.Steps
	} else {
		total = steps
	}
	fmt.Printf("\033[32mДобавлено %d шагов. Всего сегодня: %d\033[0m\n", steps, total)
}

func (c *Counter) show() {
	today := time.Now().Format("2006-01-02")
	entry := c.getToday()
	stepsToday := 0
	if entry != nil {
		stepsToday = entry.Steps
	}
	progress := stepsToday * 100 / c.data.Goal
	if progress > 100 {
		progress = 100
	}
	barLen := 20
	filled := progress * barLen / 100
	bar := ""
	for i := 0; i < filled; i++ {
		bar += "█"
	}
	for i := filled; i < barLen; i++ {
		bar += "░"
	}
	fmt.Printf("\033[36m📊 Сегодня: %d шагов\033[0m\n", stepsToday)
	fmt.Printf("\033[33mЦель: %d шагов\033[0m\n", c.data.Goal)
	fmt.Printf("Прогресс: \033[32m%s\033[0m %d%%\n", bar, progress)
	fmt.Println("\033[35mИстория за 7 дней:\033[0m")
	for i := 6; i >= 0; i-- {
		d := time.Now().AddDate(0, 0, -i).Format("2006-01-02")
		found := false
		for _, e := range c.data.History {
			if e.Date == d {
				fmt.Printf("  %s: %d\n", d, e.Steps)
				found = true
				break
			}
		}
		if !found {
			fmt.Printf("  %s: 0\n", d)
		}
	}
}

func (c *Counter) setGoal(goal int) {
	c.data.Goal = goal
	c.save()
	fmt.Printf("\033[32mЦель установлена: %d шагов\033[0m\n", goal)
}

func (c *Counter) history() {
	fmt.Println("\033[36mИстория (все записи):\033[0m")
	sort.Slice(c.data.History, func(i, j int) bool {
		return c.data.History[i].Date < c.data.History[j].Date
	})
	for _, e := range c.data.History {
		fmt.Printf("  %s: %d\n", e.Date, e.Steps)
	}
}

func (c *Counter) reset() {
	c.data.History = []HistoryEntry{}
	c.save()
	fmt.Println("\033[32mВсе данные сброшены.\033[0m")
}

func (c *Counter) exportCSV(filename string) {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка создания файла: %v\n", err)
		return
	}
	defer file.Close()
	w := csv.NewWriter(file)
	defer w.Flush()
	w.Write([]string{"date", "steps"})
	for _, e := range c.data.History {
		w.Write([]string{e.Date, strconv.Itoa(e.Steps)})
	}
	fmt.Printf("\033[32mЭкспортировано в %s (CSV)\033[0m\n", filename)
}

func main() {
	var (
		add      int
		show     bool
		goal     int
		history  bool
		reset    bool
		exportCSV string
	)
	flag.IntVar(&add, "add", 0, "Добавить шаги за сегодня")
	flag.BoolVar(&show, "show", false, "Показать сегодняшний прогресс и историю")
	flag.IntVar(&goal, "goal", 0, "Установить дневную цель")
	flag.BoolVar(&history, "history", false, "Показать всю историю")
	flag.BoolVar(&reset, "reset", false, "Сбросить все данные")
	flag.StringVar(&exportCSV, "export-csv", "", "Экспорт истории в CSV")
	flag.Parse()

	counter := &Counter{}
	counter.load()

	if add > 0 {
		counter.addSteps(add)
	} else if show {
		counter.show()
	} else if goal > 0 {
		counter.setGoal(goal)
	} else if history {
		counter.history()
	} else if reset {
		counter.reset()
	} else if exportCSV != "" {
		counter.exportCSV(exportCSV)
	} else {
		fmt.Println("Используйте --help для справки.")
	}
}

File: step_counter.rs
// step_counter.rs
use chrono::Local;
use clap::{App, Arg};
use serde::{Deserialize, Serialize};
use serde_json;
use std::fs;
use std::io::Write;
use colored::*;

const DATA_FILE: &str = "steps.json";
const DEFAULT_GOAL: i32 = 10000;

#[derive(Serialize, Deserialize)]
struct HistoryEntry {
    date: String,
    steps: i32,
}

#[derive(Serialize, Deserialize)]
struct Data {
    goal: i32,
    history: Vec<HistoryEntry>,
}

struct Counter {
    data: Data,
}

impl Counter {
    fn new() -> Self {
        let mut c = Counter { data: Data { goal: DEFAULT_GOAL, history: Vec::new() } };
        c.load();
        c
    }

    fn load(&mut self) {
        if let Ok(json) = fs::read_to_string(DATA_FILE) {
            if let Ok(data) = serde_json::from_str(&json) {
                self.data = data;
                return;
            }
        }
        self.data = Data { goal: DEFAULT_GOAL, history: Vec::new() };
    }

    fn save(&self) {
        let json = serde_json::to_string_pretty(&self.data).unwrap();
        fs::write(DATA_FILE, json).unwrap();
    }

    fn get_today(&self) -> Option<&HistoryEntry> {
        let today = Local::now().format("%Y-%m-%d").to_string();
        self.data.history.iter().find(|e| e.date == today)
    }

    fn get_today_mut(&mut self) -> Option<&mut HistoryEntry> {
        let today = Local::now().format("%Y-%m-%d").to_string();
        self.data.history.iter_mut().find(|e| e.date == today)
    }

    fn add_steps(&mut self, steps: i32) {
        let today = Local::now().format("%Y-%m-%d").to_string();
        if let Some(entry) = self.get_today_mut() {
            entry.steps += steps;
        } else {
            self.data.history.push(HistoryEntry { date: today.clone(), steps });
        }
        self.save();
        let total = self.get_today().map(|e| e.steps).unwrap_or(steps);
        println!("{}", format!("Добавлено {} шагов. Всего сегодня: {}", steps, total).green());
    }

    fn show(&self) {
        let today = Local::now().format("%Y-%m-%d").to_string();
        let steps_today = self.get_today().map(|e| e.steps).unwrap_or(0);
        let progress = (steps_today * 100 / self.data.goal).min(100);
        let bar_len = 20;
        let filled = progress * bar_len / 100;
        let bar = "█".repeat(filled as usize) + "░".repeat((bar_len - filled) as usize);
        println!("{}", format!("📊 Сегодня: {} шагов", steps_today).cyan());
        println!("{}", format!("Цель: {} шагов", self.data.goal).yellow());
        println!("Прогресс: {} {}%", bar.green(), progress);
        println!("{}", "\nИстория за 7 дней:".magenta());
        for i in (0..7).rev() {
            let d = Local::now() - chrono::Duration::days(i);
            let date_str = d.format("%Y-%m-%d").to_string();
            let found = self.data.history.iter().find(|e| e.date == date_str);
            println!("  {}: {}", date_str, found.map(|e| e.steps).unwrap_or(0));
        }
    }

    fn set_goal(&mut self, goal: i32) {
        self.data.goal = goal;
        self.save();
        println!("{}", format!("Цель установлена: {} шагов", goal).green());
    }

    fn history(&self) {
        println!("{}", "История (все записи):".cyan());
        for entry in &self.data.history {
            println!("  {}: {}", entry.date, entry.steps);
        }
    }

    fn reset(&mut self) {
        self.data.history.clear();
        self.save();
        println!("{}", "Все данные сброшены.".green());
    }

    fn export_csv(&self, filename: &str) {
        let mut wtr = csv::Writer::from_path(filename).unwrap();
        wtr.write_record(&["date", "steps"]).unwrap();
        for entry in &self.data.history {
            wtr.write_record(&[&entry.date, &entry.steps.to_string()]).unwrap();
        }
        wtr.flush().unwrap();
        println!("{}", format!("Экспортировано в {} (CSV)", filename).green());
    }
}

fn main() {
    let matches = App::new("Step Counter")
        .arg(Arg::with_name("add").long("add").takes_value(true).help("Добавить шаги за сегодня"))
        .arg(Arg::with_name("show").long("show").help("Показать сегодняшний прогресс и историю"))
        .arg(Arg::with_name("goal").long("goal").takes_value(true).help("Установить дневную цель"))
        .arg(Arg::with_name("history").long("history").help("Показать всю историю"))
        .arg(Arg::with_name("reset").long("reset").help("Сбросить все данные"))
        .arg(Arg::with_name("export-csv").long("export-csv").takes_value(true).help("Экспорт истории в CSV"))
        .get_matches();

    let mut counter = Counter::new();

    if let Some(steps) = matches.value_of("add") {
        let steps: i32 = steps.parse().expect("Неверное число");
        counter.add_steps(steps);
    } else if matches.is_present("show") {
        counter.show();
    } else if let Some(goal) = matches.value_of("goal") {
        let goal: i32 = goal.parse().expect("Неверное число");
        counter.set_goal(goal);
    } else if matches.is_present("history") {
        counter.history();
    } else if matches.is_present("reset") {
        counter.reset();
    } else if let Some(file) = matches.value_of("export-csv") {
        counter.export_csv(file);
    } else {
        println!("Используйте --help для справки.");
    }
}
